File: Editorial/0147_Insertion_Sort_List/TypeScript/1_Insertion_Sort.ts
function insertionSortList(head: ListNode | null): ListNode | null {
    let dummy = new ListNode();
    let curr = head;
    while (curr !== null) {
        // At each iteration, we insert an element into the resulting list.
        let prev = dummy;
        // find the position to insert the current node
        while (prev.next !== null && prev.next.val <= curr.val) {
            prev = prev.next;
        }
        let next = curr.next;
        // insert the current node to the new list
        curr.next = prev.next;
        prev.next = curr;
        // moving on to the next iteration
        curr = next;
    }
    return dummy.next;
}

File: Editorial/147_Insertion_Sort_List/Go/1_Insertion_Sort.go
func insertionSortList(head *ListNode) *ListNode {
    dummy := &ListNode{}
    curr := head
    for curr != nil {
        // At each iteration, we insert an element into the resulting list.
        prev := dummy
        // find the position to insert the current node
        for prev.Next != nil && prev.Next.Val <= curr.Val {
            prev = prev.Next
        }
        next := curr.Next
        // insert the current node to the new list
        curr.Next = prev.Next
        prev.Next = curr
        // moving on to the next iteration
        curr = next
    }
    return dummy.Next
}
